archive/src/expand/walk.rs

```diff
@@ -0,0 +1,245 @@
+//! Iterator based directory traversal with unarchiving.
+
+use std::{
+    collections::VecDeque,
+    fs::{self, File},
+    io,
+    path::{Path, PathBuf},
+    sync::Arc,
+    thread,
+};
+
+use crossbeam::channel::{bounded, Sender};
+use derivative::Derivative;
+use tryvial::tryvial;
+use walkdir::{DirEntry, WalkDir};
+
+use crate::{
+    strategy::{self, List},
+    Error, Options, Recursion, Target,
+};
+
+/// A directory entry discovered by the walker.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct Entry {
+    /// The logical path, relative to the expanding root.
+    /// This is the path reported to clients.
+    logical: PathBuf,
+
+    /// The actual path on disk. This is hidden from clients as an implementation detail.
+    concrete: PathBuf,
+
+    /// The [`WalkTarget`] containing the file to which this entry points.
+    /// This is needed because `WalkTarget` cleans up its directory once it finishes walking,
+    /// but `Entry` may live beyond that walk operation.
+    /// As long as `Entry` is around, its location on disk should be accessible.
+    ///
+    /// This field is not directly used; it is merely a sentinel to keep `WalkTarget` alive.
+    #[derivative(Debug = "ignore")]
+    _target: Arc<WalkTarget>,
+}
+
+impl Entry {
+    /// Create an instance with direct ancestry.
+    /// Errors if the logical entry cannot be created.
+    #[tryvial]
+    fn direct(target: Arc<WalkTarget>, dir: &Path, file: &Path) -> Result<Self, Error> {
+        let logical = try_make_relative(dir, file)?;
+        Self {
+            _target: target,
+            logical: logical.to_owned(),
+            concrete: file.to_owned(),
+        }
+    }
+
+    /// Create an instance from a walkdir entry with derived ancestry.
+    ///
+    /// Errors if the logical entry cannot be created.
+    #[tryvial]
+    fn derived(
+        target: Arc<WalkTarget>,
+        parent: Option<&Path>,
+        dir: &Path,
+        file: &Path,
+    ) -> Result<Self, Error> {
+        let entry = Self::direct(target, dir, file)?;
+        match parent {
+            Some(parent) => Entry {
+                logical: parent.join(entry.logical),
+                ..entry
+            },
+            None => entry,
+        }
+    }
```

Comment:
See previous `tryvial` comment for explanation.
```suggestion
    fn derived(
        target: Arc<WalkTarget>,
        parent: Option<&Path>,
        dir: &Path,
        file: &Path,
    ) -> Result<Self, Error> {
        let entry = Self::direct(target, dir, file)?;
        Ok(match parent {
            Some(parent) => Entry {
                logical: parent.join(entry.logical),
                ..entry
            },
            None => entry,
        })
    }
```